File: 018/4sum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[][]}
 */
var fourSum = function (nums, target) {
    nums.sort((a, b) => a - b);
    let len = nums.length;
    let res = [];

    for (let i = 0; i < len - 3; i++) {
        if (i > 0 && nums[i] == nums[i - 1]) {
            continue;
        }
        for (let j = i + 1; j < len - 2; j++) {
            if (j > i + 1 && nums[j] == nums[j - 1]) {
                continue;
            }

            let [left, right] = [j + 1, len - 1];
            while (left < right) {
                let sum = nums[i] + nums[j] + nums[left] + nums[right];
                if (sum == target) {
                    let tmp = [nums[i], nums[j], nums[left], nums[right]];
                    // tmp.push(nums[i]);
                    // tmp.push(nums[j]);
                    // tmp.push(nums[left]);
                    // tmp.push(nums[right]);
                    res.push(tmp);
                    [left, right] = next(nums, left, right);
                } else if (sum < target) {
                    left += 1;
                } else {
                    right -= 1;
                }
            }
        }
    }
    return res;
};

// find the next suitable left and right value
const next = function (nums, left, right) {
    while (left < right) {
        if (nums[left] == nums[left + 1]) {
            left++;
        } else if (nums[right] == nums[right - 1]) {
            right--;
        } else {
            left++;
            right--;
            return [left, right];
        }
    }
    return [left, right];
};

File: 018/4sum.go
package main

import (
	"fmt"
	"sort"
)

func fourSum(nums []int, target int) [][]int {
	res := [][]int{}
	sort.Ints(nums)

	for i := 0; i < len(nums)-3; i++ {
		if i > 0 && nums[i] == nums[i-1] {
			continue
		}

		for j := i + 1; j < len(nums)-2; j++ {
			if j > i+1 && nums[j] == nums[j-1] {
				continue
			}

			left, right := j+1, len(nums)-1
			for left < right {
				s := nums[i] + nums[j] + nums[left] + nums[right]
				switch {
				case s < target:
					left++
				case s > target:
					right--
				default:
					res = append(res, []int{nums[i], nums[j], nums[left], nums[right]})
					left, right = next(nums, left, right)
				}
			}
		}
	}

	return res
}

func next(nums []int, left int, right int) (int, int) {
	for left < right {
		switch {
		case nums[left] == nums[left+1]:
			left++
		case nums[right] == nums[right-1]:
			right--
		default:
			left++
			right--
			return left, right
		}
	}
	return left, right
}

func main() {
	nums := []int{1, 0, -1, 0, -2, 2}
	target := 0
	fmt.Println(fourSum(nums, target))
}
